### ISSUE TITLE
Conflict with OneSignal – Web Push Notifications wordpress plugin

### ISSUE THREAD

**Issue by SmartandCom** [Automattic/amp-wp#5297]
## Bug Description

Since AMP plugin was updated to version 2.0.0, Onesignal plugin is now sending automatic notifications notifications for every action on a post, including when working on a private draft.

## Expected Behaviour

Onesignal should not receive information that the content has been published. This is obviously sent by the new version of the AMP plugin, it used to work correctly before with previous version of the AMP plugin.

## Acceptance criteria

* 

## Implementation brief

* 

## QA testing instructions

* 

## Demo

* 

## Changelog entry

*

**Comment by westonruter** [comment]
I assume you are in Reader mode? 

What is the content of the notifications being sent when saving a draft? Is it for the post being edited it self or for a Validated URL post type? What is the URL that the notification points the user to?

This may be due to Developer Tools now being enabled by default. When Developer Tools are enabled and an administrator edits a post, the AMP plugin performs a loopback request for the permalink of the post being edited, and then stores the validation results in an `amp_validated_url` post type. 

Note that this post type is _not_ `public`:

https://github.com/ampproject/amp-wp/blob/3e538606f336c1138ee3ef213f8e0efbf14bb4a8/includes/validation/class-amp-validated-url-post-type.php#L161-L199

Nevertheless, it appears you've encountered a bug with OneSignal or at least something you can configure on the OneSignal side. See [this logic](https://github.com/OneSignal/OneSignal-WordPress-Plugin/blob/9bddbfb37a9d0c4ed2be3cdc5b0ff00c967bf1cc/onesignal-admin.php#L642-L667) in their plugin:

```php
            /* Settings related to creating a post outside of the WordPress editor NOT displaying the OneSignal meta box
             ************************************************************************************************************/

            /* OneSignal plugin setting "Automatically send a push notification when I create a post from 3rd party plugins"
             * If set to true, send only if *publishing* a post type *post* from *something other than the default WordPress editor*.
             * The filter hooks "onesignal_exclude_post" and "onesignal_include_post" can override this behavior as long as the option to automatically send from 3rd party plugins is set.
             */
            $settings_send_notification_on_non_editor_post_publish = $onesignal_wp_settings['notification_on_post_from_plugin'];
            $additional_custom_post_types_string = str_replace(' ', '', $onesignal_wp_settings['allowed_custom_post_types']);
            $additional_custom_post_types_array = array_filter(explode(',', $additional_custom_post_types_string));
            $non_editor_post_publish_do_send_notification = $settings_send_notification_on_non_editor_post_publish &&
                                                        ($post->post_type === 'post' || in_array($post->post_type, $additional_custom_post_types_array, true)) &&
                                                        $old_status !== 'publish';
            /* ********************************************************************************************************* */

            if ($posted_from_wordpress_editor) {
                // Decide to send based on whether the checkbox "Send notification on post publish/update" is checked
                // This post may be scheduled or just submitted from the WordPress editor
                // Metadata may not be saved into post yet, so use $_POST form data if metadata not available
                $do_send_notification = ($was_posted && $onesignal_meta_box_send_notification_checked) ||
                                    (!$was_posted && $post_metadata_was_send_notification_checked);
            } else {
                // This was definitely not submitted via the WordPress editor
                // Decide to send based on whether the 3rd-party plugins setting is checked
                $do_send_notification = $non_editor_post_publish_do_send_notification;
            }
```

First, try turning off the OneSignal configuration for `notification_on_post_from_plugin`. 

Otherwise, ensure that `amp_validated_url` is not among the `allowed_custom_post_types` configuration.

Alternatively, from the AMP plugin side of things, you can try turning off AMP Developer Tools in your user profile settings:

![image](https://user-images.githubusercontent.com/134745/91645026-773e0000-e9f6-11ea-827e-64da31247859.png)

Finally, OneSignal may have a bug here. If indeed notifications are being sent for the `amp_validated_url` post type, then OneSignal should be doing a check to see if `is_post_type_viewable()`. Otherwise, users will get notifications for a post type they can't even access.

**Comment by SmartandCom** [comment]
> I assume you are in Reader mode? 

Indeed, yes

> What is the content of the notifications being sent when saving a draft? Is it for the post being edited it self or for a Validated URL post type? What is the URL that the notification points the user to?

It sends just the url of the draft, even with the temporary url if it's still the case. it's not an AMP url. 

amp_validated_url is not in the list of the allowed_custom_post_types configuration.

**Comment by westonruter** [comment]
Is the draft URL for the post being edited, or for the `amp_validated_url` post type? 

What happens if you turn off Developer Tools in your WordPress user profile? Does that fix the issue?

We'll take a look, but I believe this is a bug with the OneSignal plugin.

**Comment by SmartandCom** [comment]
> What happens if you turn off Developer Tools in your WordPress user profile? Does that fix the issue?

It fixes the issue indeed, yes

**Comment by westonruter** [comment]
Do you have the “Automatically send a push notification when I publish a post from 3rd party plugins” toggle enabled?

![image](https://user-images.githubusercontent.com/134745/91648075-fe4fa000-ea17-11ea-8e8a-2306a5a11265.png)

**Comment by westonruter** [comment]
@SmartandCom I've opened a PR to fix the underlying problem in OneSignal: https://github.com/OneSignal/OneSignal-WordPress-Plugin/pull/251

If you can try to apply that change on your copy of OneSignal, and re-test with Developer Tools turned back on, that would be helpful.

**Comment by SmartandCom** [comment]
> Do you have the “Automatically send a push notification when I publish a post from 3rd party plugins” toggle enabled?

No

**Comment by SmartandCom** [comment]
If you can try to apply that change on your copy of OneSignal, and re-test with Developer Tools turned back on, that would be helpful.
---> added the code from the commit to the file
--> reactived admin tools in user profile
--> created a draft (send notification on post publish is ticked)
---> saved the draft
---> no notification is issued

So, it works

**Comment by SmartandCom** [comment]
Further testing confirms it works with your modification.
Thanks for that 
even if it's now on Onesignal side to do their work to correct their own plugin...

**Comment by westonruter** [comment]
My pull request was merged. It will be fixed in OneSignal plugin v2.1.4 per https://github.com/OneSignal/OneSignal-WordPress-Plugin/pull/251#issuecomment-683927892.
